File: solutions/typescript/game-of-life/1/game-of-life.ts
export class GameOfLife {
  private grid: number[][];
  private rows: number;
  private cols: number;

  constructor(matrix: number[][]) {
    // Create a deep copy of the matrix to avoid mutations
    this.grid = matrix.map(row => [...row]);
    this.rows = matrix.length;
    this.cols = this.rows > 0 ? matrix[0].length : 0;
  }

  private countLiveNeighbors(row: number, col: number): number {
    let liveCount = 0;
    
    // Check all 8 neighboring cells
    for (let i = -1; i <= 1; i++) {
      for (let j = -1; j <= 1; j++) {
        // Skip the cell itself
        if (i === 0 && j === 0) continue;
        
        const neighborRow = row + i;
        const neighborCol = col + j;
        
        // Check if neighbor is within bounds
        if (neighborRow >= 0 && neighborRow < this.rows && 
            neighborCol >= 0 && neighborCol < this.cols) {
          if (this.grid[neighborRow][neighborCol] === 1) {
            liveCount++;
          }
        }
      }
    }
    
    return liveCount;
  }

  public tick(): void {
    // Create a new grid for the next generation
    const nextGrid: number[][] = [];
    
    for (let i = 0; i < this.rows; i++) {
      nextGrid.push([]);
      for (let j = 0; j < this.cols; j++) {
        const liveNeighbors = this.countLiveNeighbors(i, j);
        const currentCell = this.grid[i][j];
        
        // Apply Game of Life rules
        if (currentCell === 1) {
          // Live cell: survives with 2 or 3 neighbors
          if (liveNeighbors === 2 || liveNeighbors === 3) {
            nextGrid[i][j] = 1;
          } else {
            nextGrid[i][j] = 0;
          }
        } else {
          // Dead cell: becomes alive with exactly 3 neighbors
          if (liveNeighbors === 3) {
            nextGrid[i][j] = 1;
          } else {
            nextGrid[i][j] = 0;
          }
        }
      }
    }
    
    // Update the grid
    this.grid = nextGrid;
  }

  public state(): number[][] {
    // Return a deep copy to prevent external mutations
    return this.grid.map(row => [...row]);
  }
}

File: solutions/java/game-of-life/1/src/main/java/GameOfLife.java
class GameOfLife {
    public int[][] tick(int[][] matrix) {
        if (matrix == null || matrix.length == 0) {
            return matrix;
        }
        
        int rows = matrix.length;
        int cols = matrix[0].length;
        int[][] nextGen = new int[rows][cols];
        
        // Direction vectors for 8 neighbors: up, down, left, right, and 4 diagonals
        int[] dx = {-1, -1, -1, 0, 0, 1, 1, 1};
        int[] dy = {-1, 0, 1, -1, 1, -1, 0, 1};
        
        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {
                int liveNeighbors = countLiveNeighbors(matrix, row, col, dx, dy);
                
                if (matrix[row][col] == 1) {
                    // Live cell: survives with 2 or 3 neighbors
                    nextGen[row][col] = (liveNeighbors == 2 || liveNeighbors == 3) ? 1 : 0;
                } else {
                    // Dead cell: becomes alive with exactly 3 neighbors
                    nextGen[row][col] = (liveNeighbors == 3) ? 1 : 0;
                }
            }
        }
        
        return nextGen;
    }
    
    private int countLiveNeighbors(int[][] matrix, int row, int col, int[] dx, int[] dy) {
        int count = 0;
        int rows = matrix.length;
        int cols = matrix[0].length;
        
        for (int i = 0; i < 8; i++) {
            int newRow = row + dx[i];
            int newCol = col + dy[i];
            
            // Check bounds before accessing
            if (newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols) {
                count += matrix[newRow][newCol];
            }
        }
        
        return count;
    }
}